keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -638,24 +640,149 @@ always dispatched immediately.  Following is how the other requests
+for `priority` is designed to do something more natural and convenient
+than have them all collide at some fixed number.
+
+The following table shows the current default non-exempt priority
+levels and a proposal for their new configuration.  For the sake of
+continuity with out-of-tree configuration objects, the proposed
+priority values follow the rule given above for the effective value
+when the priority field holds zero.
+
+| Name | Assured Shares | Proposed Borrowable Percent | Proposed Priority |
+| ---- | -------------: | --------------------------: | ----------------: |
+| leader-election |  10 |   0 |   150 |
+| node-high       |  40 |  25 |   400 |
+| system          |  30 |  33 |   500 |
+| workload-high   |  40 |  50 |   833 |
+| workload-low    | 100 |  90 |  9000 |
+| global-default  |  20 |  50 |  9900 |
+| catch-all       |   5 |   0 | 10000 |
+
+Each priority level has two concurrency limits: its
+NominalConcurrencyLimit (NCL) as defined above by configuration, and a
+CurrentConcurrencyLimit (CCL) that is used in dispatching requests.
+The CCLs are adjusted periodically, based on configuration, the
+current situation at adjustment time, and recent observations.  The
+"borrowing" resides in the differences between CCL and NCL.  A
+priority level's CCL can go as low as NCL-BCL; the upper limit is
+imposed only by how many seats are available for borrowing from other
+priority levels.  The sum of the CCLs, like the sum of the NCLs, is
+always equal to the server's concurrency limit (SCL).  These CCLs are
+floating-point values, because the adjustment logic below is
+incremental.  The actual limits used in dispatching are the result of
+rounding these floating-point numbers to their nearest integer.
 
 Dispatching is done independently for each priority level.  Whenever
-(1) a non-exempt priority level's number of running requests is zero
-or below the level's assured concurrency value and (2) that priority
-level has a non-empty queue, it is time to dispatch another request
-for service.  The Fair Queuing for Server Requests algorithm below is
-used to pick a non-empty queue at that priority level.  Then the
-request at the head of that queue is dispatched.
+(1) a non-exempt priority level's number of occupied seats is zero or
+below the level's rounded CCL and (2) that priority level has a
+non-empty queue, it is time to dispatch another request for service.
+The Fair Queuing for Server Requests algorithm below is used to pick a
+non-empty queue at that priority level.  Then the request at the head
+of that queue is dispatched if possible.
+
+Every 10 seconds, all the CCLs are adjusted.  The adjustments take
+into account high watermarks of seat demand.  A priority level's seat
+demand is the sum of its occupied seats and the number of seats in the
+queued requests.  Each priority level has two high watermarks: a
+short-term one M1 and a long-term one M2.  During an adjustment
```

Comment:
The description of history here makes me think of this as the integral portion of a PID.  The current state being the proportional component.

Why uses maxes for historical data instead of using seat-seconds (the amount of time that a seat was occupied).

Reply:
The proportional part of a P-I-D controller is `Kp * error`.  The reason to use M2 here is so that a priority level that just got a spike in load reclaims any seats that it had lent.  Now, it would be possible to use M2 for the lower bound and an averaged seat demand to compute the pressure (`P = AvgSeatDemand - CCL`).